src/model.rs

```diff
@@ -787,88 +792,79 @@ impl Collections {
     ///  VJ:2 (Mon-Fri)                                       12:00-13:00       14:00-15:00
     ///  VJ:3 (Sat-Sun)                                       12:30-13:30       14:30-15:30
     ///
-    /// Example 4 might be a valid use case of stay-in but is not handled in the
-    /// current implementation since the validity patterns are different; this
-    /// is undefined behavior (most likely, some stay-in will be forbidden by
-    /// the default rule).
+    /// Example 4 is a valid use case of stay-in
+    /// The pickup/dropoff will be possible between VJ:1 and VJ:2/VJ:3
     pub fn enhance_pickup_dropoff(&mut self) {
-        let mut vj_idxs: Vec<Idx<VehicleJourney>> = self
-            .vehicle_journeys
+        let mut allowed_last_pick_up_vj = HashSet::<_>::new();
+        let mut allowed_first_drop_off_vj = HashSet::<_>::new();
+
+        let can_chain_without_overlap = |prev_vj: &VehicleJourney, next_vj: &VehicleJourney| {
+            let last_stop = &prev_vj.stop_times.last();
+            let first_stop = &next_vj.stop_times.first();
+            match (last_stop, first_stop) {
+                (Some(last_stop), Some(first_stop)) => {
+                    if last_stop.stop_point_idx == first_stop.stop_point_idx {
+                        // We can discard when the stop points are identicals (see
+                        // Example 1 above).
+                        false
+                    } else {
+                        let prev_cal = self.calendars.get(&prev_vj.service_id);
+                        let next_cal = self.calendars.get(&next_vj.service_id);
+                        // for the stay in to be possible the vj should have at least one date in common
+                        prev_cal.map_or(false, |prev_cal| next_cal.map_or(false, |next_cal|prev_cal.overlaps(&next_cal))) &&
+                        // The stay in is not really possible when timing overlaps
+                        // between arrival of first vehicle journey and departure of
+                        // next vehicle journey (see Example 2 above).
+                        last_stop.departure_time <= first_stop.arrival_time
+                    }
+                }
+                _ => false,
+            }
+        };
+
+        self.vehicle_journeys
             .iter()
-            // Filtering out vehicle journeys without any stop time allowing us
-            // below to access first and last stop_times safely.
-            .filter(|(_, vj)| !vj.stop_times.is_empty())
-            .map(|(idx, _)| idx)
-            .collect();
-        // Init all vehicle journeys with:
-        // - no drop-off on first stop
-        // - no pick-up on last stop
+            .filter_map(|(vj_idx, vj)| vj.block_id.clone().map(|b| (b, (vj_idx, vj))))
+            .fold(
+                HashMap::<_, Vec<(Idx<VehicleJourney>, &VehicleJourney)>>::new(),
```

Comment:
`_` or `String` is not helping much indeed.
Adding a block-id type may be nice, but I don't know how much it would add to this PR (and if we should explore crates to help typing more).
A comment (or a good name if you refactor to a for-loop) may be a good compromise?

Reply:
This could just be creating a few aliases. Here are some ideas (not necessarily all of them has to be used, and also, naming might just be wrong!).

```rust
type VehicleJourneyIndex = (Idx<VehicleJourney>, &VehicleJourney);
type BlockVehicleJourneys = Vec<VehicleJourneyIndex>;
type BlockId = String;
type Blocks = std::collections::HashMap<BlockId, BlockVehicleJourneys>;
```

I'm mostly looking for ease of maintenance and readability here, not necessarily for strong typing.

Reply:
I only added the blockId type, tell me if that's ok for you